temporalio/activity.py

```diff
@@ -24,12 +25,55 @@
     NoReturn,
     Optional,
     Tuple,
+    TypeVar,
+    overload,
 )
 
 import temporalio.api.common.v1
 import temporalio.common
 import temporalio.exceptions
 
+ActivityFunc = TypeVar("ActivityFunc", bound=Callable[..., Any])
+
+
+@overload
+def defn(fn: ActivityFunc) -> ActivityFunc:
+    ...
+
+
+@overload
+def defn(*, name: str) -> Callable[[ActivityFunc], ActivityFunc]:
+    ...
+
+
+def defn(fn: Optional[ActivityFunc] = None, *, name: Optional[str] = None):
+    """Decorator for activity functions.
+
+    Activities can be async or non-async.
+
+    Args:
+        fn: The function to decorate.
+        name: Name to use for the activity. Defaults to function ``__name__``.
+    """
+
+    def with_name(name: str, fn: ActivityFunc) -> ActivityFunc:
+        # Validate the activity
+        if not callable(fn):
+            raise TypeError("Activity is not callable")
+        elif not fn.__code__:
+            raise TypeError("Activity callable missing __code__")
+        elif fn.__code__.co_kwonlyargcount:
+            raise TypeError("Activity cannot have keyword-only arguments")
+        # Set the name
+        setattr(fn, "__temporal_activity_name", name)
+        return fn
+
+    # If name option is available, return decorator function
+    if name is not None:
```

Comment:
You should return partial if `fn is None`, otherwise `defn()` doesn't work.

Reply:
If I did that then `defn` sans parens wouldn't work. There is no overload that supports `defn()` from the type perspective, so if they are calling it, they are violating the docs/spec here.

Reply:
WDYM no overload that supports `defn()`? Both params are optional.
I see no harm in supporting `defn()`, what are you concerned about?

Reply:
Runtime failure added

Reply:
I am merging this to get started on the rest of the workflow impl, but we can revisit this if needed.

Reply:
Oh I missed the overloads there.

Not critical but really don't see a good reason not to support that..

Reply:
But I don't see a good reason to support it.